bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieStatus.java

```diff
@@ -0,0 +1,246 @@
+        for (File dir : directories) {
+            try {
+                File statusFile = new File(dir, BOOKIE_STATUS_FILENAME);
+                writeToFile(statusFile, toString());
+                success = true;
+            } catch (IOException e) {
+                LOG.warn("IOException while trying to write bookie status to directory {}." +
+                    " This is fine if not all directories are failed.", dir);
+            }
+        }
+        if(success){
+            LOG.info("Successfully persist bookie status {}", this.bookieMode);
+        } else {
+            LOG.warn("Failed to persist bookie status {}", this.bookieMode);
+        }
+        return success;
+    }
+
+    /**
+     * Write content to the file. If file does not exist, it will create one.
+     *
+     * @param file file that you want to write to
+     * @param body content to write
+     * @throws IOException
+     */
+    private void writeToFile(File file, String body) throws IOException {
+        FileOutputStream fos = new FileOutputStream(file);
+        BufferedWriter bw = null;
+        try {
+            bw = new BufferedWriter(new OutputStreamWriter(fos, UTF_8));
+            bw.write(body);
+        } finally {
+            if (bw != null) {
+                bw.close();
+            }
+            fos.close();
+        }
+    }
+
+    /**
+     * Read bookie status from the status files, and update the bookie status if read succeed.
+     * If a status file is not readable or not found, it will skip and try to read from the next file.
+     *
+     * @param directories list of directories that store the status file
+     * @return true if successfully read from file, otherwise false.
+     */
+    boolean readFromDirectories(List<File> directories) {
+        boolean success = false;
+        for (File dir : directories) {
+            File statusFile = new File(dir, BOOKIE_STATUS_FILENAME);
+            try {
+                BookieStatus status = readFromFile(statusFile);
+                if (null != status) {
+                    synchronized (status) {
+                        if (status.lastUpdateTime > this.lastUpdateTime) {
+                            this.lastUpdateTime = status.lastUpdateTime;
+                            this.layoutVersion = status.layoutVersion;
+                            this.bookieMode = status.bookieMode;
+                            success = true;
+                        }
+                    }
+                }
+            } catch (IOException e) {
+                LOG.warn("IOException while trying to read bookie status from directory {}." +
+                    " This is fine if not all directories failed.", dir);
+            } catch (IllegalArgumentException e ){
+                LOG.warn("IllegalArgumentException while trying to read bookie status from directory {}." +
+                    " This is fine if not all directories failed.", dir);
+            }
+        }
+        if (success) {
```

Comment:
Sincere doubt: what is the semantics of `success` here? It can be false because the directories are empty (were cleaned?). Or maybe it was a stale file with a stale `lastUpdateTime`? In this case, a `success == false` not necessarily means that it "failed" to retrieve status from disk, but that there was not valid status to retrieve after all. What I am missing?

Reply:
:) This is very confusing. It will be a nightmare to figure out the behavior if we have multiple disks and each has weird status, as an operator to know what exactly going on with the bookie.

Reply:
Exactly. That is why I am saying it is confusing. Take this scenario.
1. Bookie has two journal dirs.
2. bookie info persistence was able to save to dir1, and not to dir2.
case-1: Bookie restarted, and dir1 and dir2 are readable. Bookie comes up in read-only.
case-2: Bookie restarted, and dir1 is good/readable but not dir2. Bookie comes up in read-only.
case-3: Bookie restarted, and dir1 is not readable, but dir2 is. Bookie comes up in read-write mode.

We can think to bring down the entire bookie if some of the ledgerdirs are not available, but that is a big hammer. All I am saying is this going to be operational complex to know what to expect exactly. But given that this is disabled by default, I am fine with it.

Reply:
it depends. we used same pattern in other internal systems, it isn't actually complicated. it is actually very simple and useful. because before this change, there is no way for you to ensure a bookie stays in readonly state among restarts. you have to specify --readonly or configure this in configuration when starting the bookie, that actually puts more operations overheads when you operate a 500+ bookies cluster.